ark/util/serialize.ts

```diff
@@ -39,23 +39,43 @@ export type snapshot<t, depth extends 1[] = []> =
 			[k in keyof t]: snapshot<t[k], [...depth, 1]>
 		}
 
-type snapshotPrimitive<t> =
-	t extends undefined ? "(undefined)"
-	: t extends bigint ? `${t}n`
-	: t extends symbol ? `(symbol${string})`
-	: t
+type snapshotPrimitive<t> = t extends symbol ? `(symbol${string})` : t
 
 export const print = (data: unknown, indent?: number): void =>
 	console.log(printable(data, indent))
 
+const deepReplaceUndefinedAndBigInt = (data: unknown): unknown => {
```

Comment:
I think ideally we'd want to change the way these are stringified initially, then maybe we could change the JSON.parse function in those  places to accept a reviver param (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/parse#the_reviver_parameter) to translate them?

Recursing back down could be expensive and I definitely want to make sure this change doesn't affect the core `arktype` package- it should only affect attest snapshots.

Reply:
oh this is super cool! I didn't know about the reviver param, thanks!

Reply:
So attempting to use that and it does get around the issue of the serialization error from `JSON.stringify`, but we no longer get the `5n` or `undefined` without the quotes because we need to make them into a string or something else that is parse-able.

![image](https://github.com/user-attachments/assets/33d446a2-d129-474d-b96c-a2dd1b96e42d)

I have two ideas for solving this and would like some feedback

1. We write a JSON.stringify resolver that puts some some special token in the field that we then need to deep replace before writing to file. 

```ts
JSON.stringify(
	data,
	(_key, value) => {
		if (value === undefined) return"SPECIAL_UNDEFINED"
		if (typeof value === "bigint") return"SPECIAL_BIGINT"
		return value
	},
	4
)
// deep replace here before writing to file or returning.
```
fwiw I do not like this ^

2. Drop the use of JSON.stringify and write a utility fn that does the stringification with support for undefined as a literal and bigints as well. I am not entirely sure of how hard this will be, but naively, I think it's achievable 😅 

I am hoping you have a better idea than these two, but if not, I'd be happy to attempt either one. My personal preference is probably option 2.

Reply:
If you were to do 2, I believe you'd eventually get stuck needing the same special tokens anyways, no? The limitation isn't really with the encoding/decoding method, but rather with what the format can actually store.

There are essentially two options:

1. Use a prefix as you mention. There's precedent in other cases for `$ark.` so we'd probably use that.

2. Write a custom encoding for all primitives. I already have this in a method called `serializePrimitive`. The biggest change would be that it would quote all string literals, but that would mean every string written to JSON would need escape characters like `"\"foo\""`.

I have a strong intuition that my `1` would be significantly easier to read, and that your `1` would be significantly easier to implement (and faster), so we're probably stuck with that unless you have a different idea.

Reply:
Ok, so I kinda just had to see for myself how hard it would be to implement the stringifier. 

I agree that this is likely not a great idea and that prefixing will be the better approach. I'll look into that more next, but I did commit the stringify code for now in case you are curious.

What I learned is that stringifying this in the form that we want is easy, but parseing that string is QUITE hard 😅 

I am sure there is a regex guru somewhere in the world that would love this problem, but to your point of the prefix strategy being "significantly easier to read", I completely agree and will be looking into doing that going forward.